rustzx-core/src/emulator/snapshot/szx.rs

```diff
@@ -0,0 +1,479 @@
+        // If AY needs enabling and it isn't enabled already, enable it.
+        if (flags & ZXSTAYF_128AY != 0) && (!emulator.settings.ay_enabled) {
+            emulator.enable_ay(true);
+            emulator.controller.change_mixer(&emulator.settings);
+        }
+        // AY needs disabling and is enabled currently, disable it.
+        else if (flags & ZXSTAYF_128AY == 0) && (emulator.settings.ay_enabled) {
+            emulator.enable_ay(false);
+            emulator.controller.change_mixer(&emulator.settings);
+        }
+    }
+
+    if emulator.settings.ay_enabled {
+        // chCurrentRegister
+        let ay_reg = block_data[1];
+        emulator.controller.mixer.ay.select_reg(ay_reg);
+
+        // chAyRegs
+        emulator.controller.mixer.ay.set_regs(&block_data[2..]);
+    }
+}
+
+// Process ZXSTKEYB (KEYB)
+fn process_keyb_block<H: Host>(emulator: &mut Emulator<H>, block_data: &[u8]) {
+    // dwFlags
+    // ignored for now as only issue 2 is emulated
+    let _flags = u32::from_le_bytes([block_data[0], block_data[1], block_data[2], block_data[3]]);
+
+    // chKeyboardJoystick
+    let joystick = block_data[4] as u32;
+    if joystick & ZXSTKJT_KEMPSTON != 0 {
+        emulator.controller.kempston = Some(kempston::KempstonJoy::default())
+    } else {
+        emulator.controller.kempston = None;
+    }
+}
+
+// Process ZXSTMOUSE (AMXM)
+fn process_amxm_block<H: Host>(emulator: &mut Emulator<H>, block_data: &[u8]) {
+    // chType
+    // Only Kempston mouse is supported
+    let mouse = block_data[0] as u32;
+    if mouse != 0 {
+        if mouse & ZXSTM_KEMPSTON > 0 {
+            emulator.controller.mouse = Some(KempstonMouse::default());
+        } else {
+            emulator.controller.mouse = None;
+        }
+    } else {
+        emulator.controller.mouse = None;
+    }
+}
+
+// Process ZXSTRAMPAGE (RAMP)
+fn process_ramp_block<H: Host>(
+    emulator: &mut Emulator<H>,
+    machine_id: u32,
+    block_data: &[u8],
+) -> Result<()> {
+    // wFlags
+    let flags = u16::from_le_bytes([block_data[0], block_data[1]]) as u32;
+
+    // chPageNo
+    let mut page_num = block_data[2];
+    // Remap page numbers for 16k/48k machines
+    if machine_id < ZxType::Zxstmid128k as u32 {
+        page_num = match page_num {
+            5 => 0,
+            2 => 1,
+            0 => 2,
+            _ => page_num,
+        };
+    }
+
+    let page_data = emulator.controller.memory.ram_page_data_mut(page_num);
+
+    if flags & ZXSTRF_COMPRESSED != 0 {
+        if cfg!(not(feature = "zlib")) {
+            //eprintln!("zlib decompression requires zlib feature to be enabled!");
+            return Err(SnapshotLoadError::ZlibNotSupported.into());
+        }
+        #[cfg(feature = "zlib")]
+        {
+            //let compressed_size = block_data[3..].len();
+            let compressed_data: Vec<u8> = block_data[3..].to_vec();
+            let data = decode_zlib_stream(compressed_data).unwrap();
```

Comment:
Non-trivial unwraps are not allowed in the core library, error should be thrown instead